### DISCUSSION TITLE
Error 400 when uploading multiple files using the create method

### DISCUSSION THREAD

**Discussion by axadrn** [pocketbase/pocketbase#3548] [Q&A] [answered]
### Description:
I am encountering a `400` error with the message "Failed to load the submitted data due to invalid formatting." when attempting to upload multiple files simultaneously using the `create` method. Interestingly, this issue does not occur when uploading a single file. Each of the files can be uploaded individually without any problem, but as soon as more than one file is included in a request, the error occurs.

I am utilizing `Promise.all` to handle the asynchronous upload of multiple files.

### Steps to Reproduce:
1. Add multiple files to `FormData`.
2. Call the `create` method, passing the `FormData`.
3. Observe the `400` error with the message "Failed to load the submitted data due to invalid formatting.".

### Expected Behavior:
It should be possible to upload multiple files in a single request, and the `create` method should process and store all files correctly in the database.

### Actual Behavior:
The `create` method throws a `400` error when attempting to upload multiple files in a request.

### Code Snippet:
```javascript
import { error, redirect } from '@sveltejs/kit';

interface FileObject {
  name: string;
  [key: string]: any;
}

function getFilenameWithoutExtension(file: FileObject): string {
  return file.name.split('.').slice(0, -1).join('.');
}

export const load = ({ locals }) => {
  if (!locals.pb.authStore.isValid) {
    throw redirect(303, '/login');
  }
};

export const actions = {
  create: async ({ request, locals }) => {
    const body = await request.formData();
    const files: FileObject[] = body.getAll('file');

    if (files.length === 0) {
      throw error(400, 'No files provided');
    }

    const promises = files.map(async (file, index) => {
      if (file.size === 0) {
        return;  // Skip empty files
      }

      const nameWithoutExtension = getFilenameWithoutExtension(file);

      const formData = new FormData();
      formData.append('user', locals.user.id);
      formData.append('name', nameWithoutExtension);
      formData.append('file', file);

      try {
        await locals.pb.collection('documents').create(formData, { requestKey: null });
      } catch (err) {
        console.log('Error: ', err);
        throw error(err.status, err.message);
      }
    });

    try {
      await Promise.all(promises);  // Utilizing Promise.all here
    } catch (err) {
      console.log('Error: ', err);
      throw error(500, 'Failed to create documents');
    }

    throw redirect(303, '/documents');
  }
};
```

### Error Message:
```bash
Error:  ClientResponseError 400: Failed to load the submitted data due to invalid formatting.
    at file:///Users/xxx/Coding/pocket/web/node_modules/pocketbase/dist/pocketbase.es.mjs:1:34068
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async eval (/Users/xxx/Coding/pocket/web/src/routes/(app)/documents/new/+page.server.ts:31:9)
    at async Promise.all (index 1)
    at async create (/Users/xxx/Coding/pocket/web/src/routes/(app)/documents/new/+page.server.ts:38:7)
    at async Module.handle_action_json_request (/Users/xxx/Coding/pocket/web/node_modules/@sveltejs/kit/src/runtime/server/page/actions.js:57:16)
    at async resolve (/Users/xxx/Coding/pocket/web/node_modules/@sveltejs/kit/src/runtime/server/respond.js:415:18)
    at async Object.handle (/Users/xxx/Coding/pocket/web/src/hooks.server.ts:17:20)
    at async Module.respond (/Users/xxx/Coding/pocket/web/node_modules/@sveltejs/kit/src/runtime/server/respond.js:282:20)
    at async file:///Users/xxx/Coding/pocket/web/node_modules/@sveltejs/kit/src/exports/vite/dev/index.js:510:22 {
  url: 'http://localhost:8090/api/collections/documents/records',
  status: 400,
  response: {
    code: 400,
    message: 'Failed to load the submitted data due to invalid formatting.',
    data: {}
  },
  isAbort: false,
  originalError: {
    url: 'http://localhost:8090/api/collections/documents/records',
    status: 400,
    data: {
      code: 400,
      message: 'Failed to load the submitted data due to invalid formatting.',
      data: {}
    }
  }
}
Error:  HttpError {
  status: 400,
  body: {
    message: 'Failed to load the submitted data due to invalid formatting.'
  }
}
```

### Environment:
- PocketBase Version: (v0.18.8)
- Node.js Version: (v20.5.1)
- Browser: (Safari, Firefox)

### Additional Remarks:
I can provide further information or debugging details if needed. The emphasis on utilizing Promise.all is to ensure all file uploads are completed before proceeding, yet the error occurs when more than one file is included in the request.

**Comment by ganigeorgiev** [comment]
I don't think this is an issue with PocketBase or the JS SDK as I'm not able to reproduce it locally with the static JS example from https://pocketbase.io/docs/files-handling/#uploading-files.

You can also start PocketBase with `--debug` and check if there is a more detailed error printed in the console.

I suspect that it is something with the `FileObject` extraction. I don't know what that is, but note that the `FormData` value can be either a string or `File`/`Blob` instance so double check that `body.getAll('file')` returns the expected value. 

The `map` Promise generation is also kindof strange (_~since you catch the request error in the map I don't think `Promise.all` will ever return an error, unless some of the other function throws~ actually you rethrow the error_) and I'd recommend to either return directly the request promise or convert it to a plain `for` loop and push the request promise manually, something like:

```js
const promises = [];

for (let file of files) {
  if (file.size === 0) {
    continue;
  }

  const nameWithoutExtension = getFilenameWithoutExtension(file);

  const formData = new FormData();
  formData.append('user', locals.user.id);
  formData.append('name', nameWithoutExtension);
  formData.append('file', file);
  promise.push(locals.pb.collection('documents').create(formData, { requestKey: null }));
}

// use allSettled to run all promises even on rejection
await Promise.allSettled(promises);
```

If you are not able to identify the cause of the issue, please create a **minimal** reproducible repo and I'll have a look.

**Reply by ganigeorgiev** [reply]
Another thing that can cause the above error I guess is incompatible `fetch`/`FormData` polyfill.

If you use the SvelteKit node adapter, try to disable the `polyfill` option - https://kit.svelte.dev/docs/adapter-node#options to use the native implementations.

**Reply by axadrn** [reply]
Thank you very much for your support and the suggested solutions thus far. I have implemented the recommended changes by using the for loop instead of the map method and disabling the polyfills, but unfortunately, the issue persists.

I have uploaded a sample project on GitHub to demonstrate the issue:

[https://github.com/DerbeDotDev/pb-bug-example](https://github.com/DerbeDotDev/pb-bug-example)

Interestingly, the Client-Side Rendering (CSR) solution works without any problems, whereas the Server-Side Rendering (SSR) solution does not, which is quite unusual.

I've noticed that the SSR solution works fine when only one file is selected for upload; however, the error occurs 99% of the time when two or more files are selected. Strangely, in about 1% of my tests, the SSR solution did work even with multiple files, but this success was not consistent.  Additionally, it's worth noting that there can be a delay of 1-2 minutes before the error message appears.

I am uncertain whether the issue lies with SvelteKit or PocketBase, and the error continues to occur even after implementing the suggested solutions. The error message remains as:

```
ClientResponseError 400: Failed to load the submitted data due to invalid formatting.
    at file:///Users/xxx/Coding/my-app/node_modules/pocketbase/dist/pocketbase.es.mjs:1:31144
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Promise.all (index 1)
    at async create (/Users/xxx/Coding/my-app/src/routes/ssr/+page.server.ts:25:7)
    at async Module.handle_action_json_request (/Users/xxx/Coding/my-app/node_modules/@sveltejs/kit/src/runtime/server/page/actions.js:57:16)
    at async resolve (/Users/xxx/Coding/my-app/node_modules/@sveltejs/kit/src/runtime/server/respond.js:415:18)
    at async Module.respond (/Users/xxx/Coding/my-app/node_modules/@sveltejs/kit/src/runtime/server/respond.js:282:20)
    at async file:///Users/xxx/Coding/my-app/node_modules/@sveltejs/kit/src/exports/vite/dev/index.js:510:22 {
  url: 'http://127.0.0.1:8090/api/collections/documents/records',
  status: 400,
  response: {
    code: 400,
    message: 'Failed to load the submitted data due to invalid formatting.',
    data: {}
  },
  isAbort: false,
  originalError: {
    url: 'http://127.0.0.1:8090/api/collections/documents/records',
    status: 400,
    data: {
      code: 400,
      message: 'Failed to load the submitted data due to invalid formatting.',
      data: {}
    }
  }
}
```

I have also attached a screenshot of the error in the Pocketbase Admin Dashboard.
<img width="1512" alt="Screenshot 2023-10-19 at 11 42 47" src="https://github.com/pocketbase/pocketbase/assets/26936893/b46216ea-205f-44de-9887-85f215c98f86">


I would greatly appreciate any further suggestions or support, and am available to provide additional information if needed.

Thank you for your time and assistance with this matter.

**Reply by ganigeorgiev** [reply]
I'll check the repository sometime later today, but I don't think this is a PocketBase issue.

The `i/o timeout` error message indicates that for some reason the upload request was interrupted. It could be a node server error, reverse proxy or if you use Docker something related to its network configuration.
If you have large files that takes longer time to upload then this is also possible to be the cause because by default we have a 10min timeout.

You've already confirmed that the error is not reproducible client-side, but to rule out a PocketBase issue, you can create a static HTML file in `pb_public/index.html`:

```html
<!DOCTYPE html />
<html>
    <head>
        <meta charset="utf-8" />
        <title>Test create</title>
    </head>
    <body>
        <input type="file" name="file" id="file" multiple>
        <button type="submit" id="btn">Submit</button>

        <script type="module">
            import PocketBase from "https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@0.18.3/dist/pocketbase.es.mjs"

            const pb = new PocketBase("http://127.0.0.1:8090");

            const fileInput = document.getElementById("file")

            document.getElementById("btn").addEventListener("click", async (e) => {
                const promises = [];

                for (let i = 0; i < 100; i++) {
                    promises.push(pb.collection("example").create({
                        file:  Array.from(fileInput.files),
                    }, { requestKey: null }));
                }

                await Promise.all(promises);
            });
        </script>
    </body>
</html>
```

The above expect a public "example" collection with one multiple "file" `file` field.
When you access `http:127.0.0.1:8090` in your browser, select some files and click "Submit", it will create 100 records in the "example" collection.

If the above works, then the problem is definitely somewhere in the node/env setup.

**Reply by ganigeorgiev** [reply]
I've tested your minimal repo (with `npm run dev`) and it works fine for me in both cases uploading 3 images each ~300kb (_the CSR example has a warning that `use:enchance` can be used only with `method="POST"`_).

The cause of the issue is something in your node/env setup but I'm not sure how to help you debug it.
If you use Docker and/or have a reverse proxy (nginx, apache, caddy, etc.) double check their configurations and body size limits.

**Reply by ganigeorgiev** [reply, marked as answer]
@DerbeDotDev Actually I was able to reproduce the issue with different Node.js version.

It doesn't seem to work with `node v20.5.0` but if you downgrade to `node v18.x` or upgrade to `node v21.0.0` then it works fine.

It seems like an issue with node/node-fetch implementation for that specific `v20.x` release (_I'm not sure what exactly is causing it as I cannot quickly find a related issue but upgrading seems to work fine_).
Try to upgrade to `node v21.0.0` and try again (you can install `nvm` and run `nvm use 21` for example to test with different versions).

**Comment by axadrn** [comment]
> @DerbeDotDev Actually I was able to reproduce the issue with different Node.js version.
> 
> It doesn't seem to work with `node v20.5.0` but if you downgrade to `node v18.x` or upgrade to `node v21.0.0` then it works fine.
> 
> It seems like an issue with node/node-fetch implementation for that specific `v20.x` release (_I'm not sure what exactly is causing it as I cannot quickly find a related issue but upgrading seems to work fine_). Try to upgrade to `node v21.0.0` and try again (you can install `nvm` and run `nvm use 21` for example to test with different versions).


Thank you immensely for your assistance and thorough investigation into this issue. I have updated Node from version 20.5.x to 20.8.x, and now everything is functioning correctly. I was nearing my wit's end, and I can't express enough how grateful I am for your help. The issue with the Node version would likely have been one of the last things I would have thought to check.

It might be beneficial to note this Node version-related issue in the PocketBase FAQ or documentation to assist others who might encounter a similar problem in the future.

Thank you once again for your time, effort, and everything you've done to help resolve this issue. Your support has been invaluable, and I greatly appreciate it.
